### PR TITLE
[MM-387] Add a global welcome message field

### LINKED ISSUE #36: allow messages for *all* teams
currently we can specify a welcome-text for each team separately.

however, i would like to be able to send people a welcome-message after their first login - regardless of the team they join.

Probably the simplest solution (from a admin-pov) is to be able to omit the `TeamName` or use a wild-card.

E.g. the following would remind a user of the code-of-conduct the first time they log-in (the `TeamName` field has been omitted), and gives a welcome-message for each team they join (the `TeamName` is `*`):

~~~yml
"Plugins": {
    "com.mattermost.welcomebot": {
        "WelcomeMessages": [
            {
                "DelayInSeconds": 1,
                "Message": [
                    "### Welcome {{.UserDisplayName}}!",
                    "",
                    "Please check out our [Code-Of-Conduct](https://mm.example.com/coc/) before using MatterMost",
                ]
            },
            {
                "TeamName": "*",
                "DelayInSeconds": 5,
                "Message": [
                    "### Welcome {{.UserDisplayName}} to the {{.Team.DisplayName}} team!",
                    "",
                    "Feel free to discuss off-topic things, but please use the ~off-topic channel for this",
                ]
            }
    ]
  }
}
~~~

### PR DESCRIPTION
**Pull request by ayusht2810**
#### Summary
- Added a global welcome message field
- Updated Readme with the addition of new field

#### Screenshot
image

#### Checklist
- [x] Completed dev testing
- [x] `make test` Ran test cases and ensured they are passing
- [x] `make check-style` Ran style check and ensured both webapp and server pass the checks

#### Ticket Link
Fixes #36 

#### What to test?
- Add a new field `GlobalWelcomeMessage` in the welcome bot config
- Create a new user and log in the Mattermost server from the new user
- Check for the global message present in the welcome bot DM
- Some edge cases to test: Add only `GlobalWelcomeMessage` in the config, add both `TeamName` and `GlobalWelcomeMessage` in the config

### CODE REVIEW

File: server/welcomebot.go [hunk 1]
```diff
@@ -45,6 +45,29 @@ func (p *Plugin) constructMessageTemplate(userID, teamID string) *MessageTemplat
 	return data
 }
 
+func (p *Plugin) constructGlobalMessageTemplate(userID string) *GloablMessageTemplate {
```
Comment (raghavaggarwal2308) [inline server/welcomebot.go:48]:
```suggestion
func (p *Plugin) getGlobalMessageTemplateData(userID string) *GloablMessageTemplate {
```
I think we should rename this

File: server/hooks.go [hunk 2]
```diff
@@ -1,14 +1,52 @@
...
+		return
+	}
+
+	for _, message := range p.getWelcomeMessages() {
+		if data.User.IsGuest() && !message.IncludeGuests {
+			continue
+		}
+
+		if len(message.GlobalWelcomeMessage) > 0 {
+			tmpMsg, _ := template.New("Response").Parse(strings.Join(message.GlobalWelcomeMessage, "\n"))
```
Comment (raghavaggarwal2308) [inline server/hooks.go:28]: Can create a constant for `Response`

**Review by raghavaggarwal2308 — CHANGES_REQUESTED** [review]

**Review by raghavaggarwal2308 — APPROVED** [review]

File: server/message_template.go [hunk 3]
```diff
@@ -11,3 +11,11 @@ type MessageTemplate struct {
 	DirectMessage   *model.Channel
 	UserDisplayName string
 }
+
+// GloablMessageTemplate represents all the data that can be used in the template for a welcomebot global message
+type GloablMessageTemplate struct {
```
Comment (mickmister) [inline server/message_template.go:16]:
```suggestion
// GlobalMessageTemplate represents all the data that can be used in the template for a welcomebot global message
type GlobalMessageTemplate struct {
```

File: server/welcomebot.go [hunk 4]
```diff
@@ -45,6 +45,29 @@ func (p *Plugin) constructMessageTemplate(userID, teamID string) *MessageTemplat
 	return data
 }
 
+func (p *Plugin) getGlobalMessageTemplateData(userID string) *GloablMessageTemplate {
```
Comment (mickmister) [inline server/welcomebot.go:48]: Can we have this function return an error as a second return value?

File: server/welcomebot.go [hunk 5]
```diff
@@ -45,6 +45,29 @@ func (p *Plugin) constructMessageTemplate(userID, teamID string) *MessageTemplat
...
+	data := &GloablMessageTemplate{}
+	var err *model.AppError
```
Comment (mickmister) [inline server/welcomebot.go:50]: `appErr` is usually used for this to make this obvious

```suggestion
	var appErr *model.AppError
```

File: server/hooks.go [hunk 6]
```diff
@@ -1,14 +1,52 @@
...
+	if data == nil {
+		return
+	}
+
+	for _, message := range p.getWelcomeMessages() {
+		if data.User.IsGuest() && !message.IncludeGuests {
+			continue
+		}
+
+		if len(message.GlobalWelcomeMessage) > 0 {
```
Comment (mickmister) [inline server/hooks.go:27]: To avoid indentation

```suggestion
		if len(message.GlobalWelcomeMessage) == 0 {
		    continue
		}
```

File: server/hooks.go [hunk 7]
```diff
@@ -1,14 +1,52 @@
...
+				Message:   message.String(),
+				UserId:    p.botUserID,
+				ChannelId: data.DirectMessage.Id,
+			}
+
+			if _, err := p.API.CreatePost(post); err != nil {
+				p.API.LogError("We could not create the response post", "UserID", post.UserId, "Error", err.Error())
+			}
+
+			return
```
Comment (mickmister) [inline server/hooks.go:45]: I'm not sure if we should return here. What if there are more messages?

**Review by mickmister — COMMENTED** [review]
Overall logic LGTM, just a few suggestions

**Comment by ayusht2810** [comment]
@mickmister Fixed the review fixes. Please re-review
